examples/service/service.1.ronn
lomiri-location-service(1) -- A location service aggregating position/velocity/heading updates and exporting them over dbus.
============================================================================================================================

## SYNOPSIS
`lomiri-location-service` `--bus` <bus> `--provider` <provider> [`--provider` <provider>] [provider specific options] [logging options]

## DESCRIPTION
A location service aggregating position/velocity/heading updates and exporting them over dbus.

## OPTIONS
These options control which bus the service is exposed upon and which providers are instantiated:

  * `--bus`=<bus>:
  Specify the bus that the service should be exposed upon, valid values are "session" or "system".
  
  * `--provider`=<provider>:
  Specify the provider that should be instantiated and configured for this service instance.
  Valid values are: geoclue::Provider, gps::Provider and skyhook::Provider.
    
These options control the behavior of the geoclue provider:

  * `--geoclue::Provider::name`=<name>:
  The name of the provider on the bus.
  
  * `--geoclue::Provider::path`=<path>:
  The object path of the provider on the bus.
  
These options control the behavior of the Skyhook provider:

  * `--skyhook::Provider::username`=<username>:
  Username for authenticating with the Skyhook service.
  
  * `--skyhook::Provider::realm`=<realm>:
  Realm for authenticating with the Skyhook service.
  
  * `--skyhook::Provider::period`=<period>:
  Delay between updates in [ms].
  
These options control the logging behavior of the service:

  * `--logtostderr`=<boolean, default: false>:
  Log messages to stderr instead of logfiles, defaults to false.
  
  * `--stderrthreshold`=<threshold, default: 2>:
  Copy log messages at or above this level to stderr in addition to logfiles. The numbers of severity levels INFO, WARNING, ERROR, and FATAL are 0, 1, 2, and 3, respectively. Defa

  * `--minloglevel`=<level, default: 0>:
  Log messages at or above this level. Again, the numbers of severity levels INFO, WARNING, ERROR, and FATAL are 0, 1, 2, and 3, respectively.
  
  * `--log_dir`=<path, default: "">:
  If specified, logfiles are written into this directory instead of the default logging directory.

  * `--v`=<level, default: 0>:
  Show all VLOG(m) messages for m less or equal the value of this flag.

## EXAMPLES

Starting the service with a gps and a skyhook provider, exposing it on the system bus:
    $ /usr/bin/lomiri-location-service/examples/service
        --bus system 
        --provider gps::Provider 
        --provider skyhook::Provider 
        --skyhook::Provider::username=a_registered_user_name 
        --skyhook::Provider::realm=a_registered_realm

A slightly more verbose instance:
    $ /usr/bin/lomiri-location-service/examples/service      
        --bus system
        --provider gps::Provider 
        --provider skyhook::Provider 
        --skyhook::Provider::username=a_registered_user_name
        --skyhook::Provider::realm=a_registered_realm
        --logtostderr=true
        --v=5

## COPYRIGHT

Lomiri Location Service is Copyright (C) 2013 Canonical Ltd.


ChangeLog
2025-02-07 Mike Gabriel

        * Release 3.3.0 (HEAD -> main, tag: 3.3.0)

2025-02-07 Ratchanan Srirattanamet

        * Merge branch 'personal/gberh/optional-trust-store' into 'main'
          (a54c75d)

2025-02-05 Guido Berhoerster

        * Add compile-time option for trust-store support (74c0e46)

2025-01-12 தமிழ்நேரம்

        * Translated using Weblate (Tamil) (1118f26)

2024-12-28 Hosted Weblate

        * Update translation files (e762442)

2024-12-22 Hosted Weblate

        * Update translation files (ea3a35d)

2024-12-22 Mike Gabriel

        * Release 3.2.0 (cf3527a) (tag: 3.2.0)

2024-12-22 Marius Gripsgard

        * Merge branch 'personal/sunweaver/fully-rename' into 'main'
          (d5e0f3c)

2024-12-21 Mike Gabriel

        * po/: Modernize i18n. (713c600)
        * debian/: Fully rename src:pkg from location-service to
          lomiri-location-service. (18ba981)
        * d/control: Drop lomiri-location-service-doc from R:, already in S:.
          (ffc7ced)
        * d/copyright: No duplicate Copyright: field in second stanza.
          (d990212)
        * d/copyright: Move global globbing license stanza to the top.
          (6c1881d)
        * d/lomiri-location-service-doc.install: Adjust to new target dir for
          doc files. (17887de)
        * doc/CMakeLists.txt: Put doc files into
          /usr/share/doc/<project-name>. (bbbdf3c)
        * CMakeLists.txt: Bump to 3.2.0. Add LANGUAGES parameter to project()
          macro. (ee25662)

2024-12-21 Marius Gripsgard

        * Merge branch 'personal/peat-psuwit/fix-etc-gps.conf' into 'main'
          (f0ef972)

2024-12-21 Ratchanan Srirattanamet

        * d/rules: fix the way /etc/gps.conf is installed (dc87f5a)

2024-12-20 Mike Gabriel

        * Merge branch 'personal/mariogrip/fixgps' into 'main' (73c1ebf)

2024-12-20 Marius Gripsgard

        * debian: make sure gps.conf ends up in etc (2480df8)

2024-12-20 Mike Gabriel

        * debian/{liblomiri-location-service3.install,rules}: Install
          gps.conf only if LOCATION_SERVICE_ENABLE_GPS_PROVIDER is
          set to ON. (bc60a38)
        * debian/rules: Make dh_auto_configure call more readable. (e0401aa)

2024-12-20 Marius Gripsgard

        * debian: Make gps.conf optional (1467206)

2024-12-14 Marius Gripsgard

        * debian: Only enable gps provider on ubuntu (0edaee8)

2024-12-13 Marius Gripsgard

        * cmake: Remove prepush as this is bzr and debian spesific and not
          used (9fff715)
        * 3rd-party: Use correct import path for newer boost (c6c9a2e)

2024-12-16 Distant Prince

        * Translated using Weblate (English (Canada)) (219fe85)

2024-12-13 Marius Gripsgard

        * Merge branch 'remove-hardware-alarm-dependency-1' into 'main'
          (a152c3d)

2024-10-25 Alexander Richards

        * com/lomiri/location: remove hardware alarm use for time_since_boot
          (677b49b)

2024-08-16 Jack Torrent

        * Translated using Weblate (Turkish) (f614403)

2024-08-05 umesaburo sagawa

        * Translated using Weblate (Japanese) (52593b8)

2024-05-16 Daniel Soohan Park

        * Translated using Weblate (Korean) (1605d7f)

2024-03-26 naizena

        * Translated using Weblate (Basque) (808a491)

2024-01-27 Mike Gabriel

        * CMakeLists.txt: Set PROJECT_VERSION in project() macro, marks
          'other' version as library SOVERSION / API version.
          (c093828)
        * CMakeLists.txt: Bump cmake_minimum_required() to version 3.5.
          (d196e8a)

2023-12-27 Weblate

        * Added translation using Weblate (Chinese (Simplified)
          (zh_LATN@pinyin)) (6da951a)
        * Added translation using Weblate (Tamil (Sri Lanka)) (358d711)
        * Added translation using Weblate (Meadow Mari) (42b81cc)
        * Added translation using Weblate (Crimean Tatar) (8dc137b)
        * Added translation using Weblate (Chuvash) (2b7a081)
        * Added translation using Weblate (Silesian) (fd8cdf9)
        * Added translation using Weblate (Venetian) (861efd6)
        * Added translation using Weblate (Ossetian) (acaa497)
        * Added translation using Weblate (Sami (Northern)) (5be95c5)
        * Added translation using Weblate (Greenlandic) (c888fed)
        * Added translation using Weblate (Assamese) (60e8b9c)
        * Added translation using Weblate (Haitian) (8fe795a)
        * Added translation using Weblate (Cornish) (3f93139)
        * Added translation using Weblate (Frisian) (51e8295)
        * Added translation using Weblate (Kashubian) (1812733)
        * Added translation using Weblate (Aragonese) (e692b07)
        * Added translation using Weblate (Tibetan) (72fe048)
        * Added translation using Weblate (Tigrinya) (50567d0)

2023-12-26 Weblate

        * Added translation using Weblate (Tatar) (a16afa8)
        * Added translation using Weblate (German (Walser)) (0abee56)
        * Added translation using Weblate (Italian (it_CARES)) (2aa69af)
        * Added translation using Weblate (Latin) (50ecb70)
        * Added translation using Weblate (Filipino) (3e6a7da)

2023-12-04 Reza Almanda

        * Translated using Weblate (Indonesian) (1788c3a)

2023-11-10 Marius Gripsgard

        * Merge branch 'missing-headers' into 'main' (4900473)

2023-11-04 Brandon Boese

        * Add cstdint to location includes (38806f7)

2023-10-18 Weblate

        * Added translation using Weblate (Xhosa) (37f17c4)
        * Added translation using Weblate (Manx) (93fb3ba)
        * Added translation using Weblate (Chechen) (b632708)
        * Added translation using Weblate (Wolof) (fe192f3)
        * Added translation using Weblate (Uzbek) (04c95a6)
        * Added translation using Weblate (Kyrgyz) (cff99e6)
        * Added translation using Weblate (Kurdish) (954fa15)
        * Added translation using Weblate (Sotho (Southern)) (4c7d429)
        * Added translation using Weblate (Norwegian Nynorsk) (8d16a52)
        * Added translation using Weblate (Swahili) (77922c3)

examples/service/service.cpp
/*
 * Copyright © 2012-2013 Canonical Ltd.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Authored by: Thomas Voß <thomas.voss@canonical.com>
 */
#include "program_options.h"

#include <com/lomiri/location/provider_factory.h>
#include <com/lomiri/location/boost_ptree_settings.h>

#include <com/lomiri/location/service/default_configuration.h>
#include <com/lomiri/location/service/implementation.h>

#include <core/dbus/announcer.h>
#include <core/dbus/asio/executor.h>

#include <thread>

namespace cll = com::lomiri::location;
namespace clls = com::lomiri::location::service;
namespace dbus = core::dbus;

namespace
{
struct NullReporter : public clls::Harvester::Reporter
{
    NullReporter() = default;

    /** @brief Tell the reporter that it should start operating. */
    void start() override
    {
    }

    /** @brief Tell the reporter to shut down its operation. */
    void stop()
    {
    }

    /**
     * @brief Triggers the reporter to send off the information.
     */
    void report(const cll::Update<cll::Position>&,
                const std::vector<cll::connectivity::WirelessNetwork::Ptr>&,
                const std::vector<cll::connectivity::RadioCell::Ptr>&)
    {
    }
};
}

int main(int argc, char** argv)
{
    cll::ProgramOptions options;

    options.add("help", "Produces this help message");
    options.add(
        "bus", 
        "The well-known bus to announce the service upon", 
        std::string{"session"});
    options.add(
        "config-file",
        "The configuration we should read from/write to",
        std::string{"/var/run/lomiri-location-service/config.ini"});
    options.add_composed<std::vector<std::string>>(
        "provider", 
        "The providers that should be added to the engine");

    if (!options.parse_from_command_line_args(argc, argv))
        return EXIT_FAILURE;

    if (options.value_count_for_key("help") > 0)
    {
        options.print_help(std::cout);
        return EXIT_SUCCESS;
    }
    
    if (options.value_count_for_key("provider") == 0)
    {
        std::cout << "A set of providers need to be specified. The following providers are known:" << std::endl;
        cll::ProviderFactory::instance().enumerate(
            [](const std::string& name, const cll::ProviderFactory::Factory&)
            {
                std::cout << "\t" << name << std::endl;
            });
        return EXIT_FAILURE;
    }

    auto selected_providers = options.value_for_key<std::vector<std::string>>("provider");

    std::map<std::string, cll::ProviderFactory::Configuration> config_lut;

    clls::DefaultConfiguration config;
    auto settings = std::make_shared<cll::BoostPtreeSettings>(options.value_for_key<std::string>("config-file"));
    auto engine = config.the_engine(std::set<cll::Provider::Ptr>{}, config.the_provider_selection_policy(), settings);

    for (const std::string& provider : selected_providers)
    {
        std::cout << "Instantiating and configuring: " << provider << std::endl;
        options.enumerate_unrecognized_options(
            [&config_lut, provider](const std::string& s)
            {
                std::stringstream in(s);
                std::string key, value;
                
                std::getline(in, key, '=');
                std::getline(in, value, '=');
                
                std::size_t pos = key.find(provider);
                if (pos == std::string::npos)
                    return;
                static const std::string option_marker{"--"};
                static const std::string scope_separator{"::"};
                key = key.erase(key.find_first_of(option_marker), option_marker.size()); 
                key = key.erase(key.find_first_of(provider), provider.size());
                key = key.erase(key.find_first_of(scope_separator), scope_separator.size()); 
                
                std::cout << "\t" << key << " -> " << value << std::endl;

                config_lut[provider].put(key, value);
            });

        try
        {
            auto result = std::async(std::launch::async, [provider, config_lut, engine] {
                return cll::ProviderFactory::instance().create_provider_for_name_with_config(
                    provider,
                    config_lut.at(provider),
                    [engine](cll::Provider::Ptr provider)
                    {
                        engine->add_provider(provider);
                    });
            });
        } catch(const std::runtime_error& e)
        {
            std::cerr << "Exception instantiating provider: " << e.what() << " ... Aborting now." << std::endl;
            return EXIT_FAILURE;
        }
    }
    
    static const std::map<std::string, dbus::WellKnownBus> lut = 
    {
        {"session", dbus::WellKnownBus::session},
        {"system", dbus::WellKnownBus::system},
    };

    dbus::Bus::Ptr incoming
    {
        new dbus::Bus{lut.at(options.value_for_key<std::string>("bus"))}
    };
    incoming->install_executor(dbus::asio::make_executor(incoming));

    dbus::Bus::Ptr outgoing
    {
        new dbus::Bus{lut.at(options.value_for_key<std::string>("bus"))}
    };
    outgoing->install_executor(dbus::asio::make_executor(outgoing));

    clls::Implementation::Configuration configuration
    {
        incoming,
        outgoing,
        engine,
        config.the_permission_manager(incoming),
        clls::Harvester::Configuration
        {
            cll::connectivity::platform_default_manager(),
            std::make_shared<NullReporter>()
        }
    };

    auto location_service = std::make_shared<clls::Implementation>(configuration);
    
    std::thread t1{[incoming](){incoming->run();}};
    std::thread t2{[outgoing](){outgoing->run();}};
    
    if (t1.joinable())
        t1.join();

    if (t2.joinable())
        t2.join();

    return EXIT_SUCCESS;
}
